### ISSUE TITLE
TLSMemoryBIOProtocol instances should provide underlying transport's interfaces, in particular ITCPTransport

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5182]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5182|
|Type|release blocker: regression|
|Created|2011-07-07 12:38:01Z|
|Branch|https://github.com/twisted/twisted/tree/tls-interfaces-5182|

While `ProtocolWrapper.__getattr__` ensures ITCPTransport methods can be called, the fact that the interface is implemented is not visible on the interface introspection level. TLSMemoryBIOProtocol instances should provide ITCPTransport (and whatever other interfaces there are) if the underlying transport does.

<details><summary>Searchable metadata</summary>

```
trac-id__5182 5182
type__release_blocker__regression release blocker: regression
reporter__itamar itamar
priority__normal normal
milestone__Twisted_12_2 Twisted-12.2
branch__branches_tls_interfaces_5182 branches/tls-interfaces-5182
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1310042281000000 1310042281000000
changetime__1345234947000000 1345234947000000
version__None None
owner__itamar itamar
cc__ashwini.oruganti@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35140]) Branching to 'tls-interfaces-5182'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

I checked in an implementation, but it is incorrect; I need to switch it to the wrapped *transport*, not *protocol*, so the code has to be moved to `makeConnection` probably.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

I've started a buildbot run here:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tls-interfaces-5182

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Okay.  Looks pretty good.  Perhaps the news fragment could also make it clear that this applies to any code using listenSSL or connectSSL (with a sufficiently new version of pyOpenSSL - but without being so verbose :).  Merge when you're satisfied.

Separately, do you remember why you filed this ticket?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set status to `closed`|
|-|-|

(In [35277]) Merge tls-interfaces-5182: TLSMemoryBIOProtocol now provides underlying transport's interfaces.

Author: itamar
Review: exarkun
Fixes: [#5182](https://github.com/twisted/twisted/issues/5182)

SSL transports now provide (once again) things like ITCPTransport when wrapping TCP transports.
